mono/utils/mono-utility-thread.c

```diff
@@ -0,0 +1,192 @@
+/**
+ * \file
+ *     A lightweight worker thread with lockless messaging
+ *
+ * Author:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * (C) 2018 Microsoft, Inc.
+ *
+ */
+
+#include <mono/utils/mono-utility-thread.h>
+
+typedef struct {
+	MonoLockFreeQueueNode node;
+
+	// For cleanup metadata
+	MonoUtilityThread *thread;
+
+	// For synch calls
+	gboolean *finished;
+	MonoSemType *response_sem;
+
+	// Variably-sized, size is thread->payload_size
+	gpointer payload [MONO_ZERO_LEN_ARRAY]; 
+} UtilityThreadQueueEntry;
+
+static void
+free_queue_entry (gpointer p)
+{
+	UtilityThreadQueueEntry *util = (UtilityThreadQueueEntry *) p;
+	mono_lock_free_free (p, util->thread->message_block_size);
+}
+
+static gboolean
+utility_thread_handle_inbox (MonoUtilityThread *thread, gboolean at_shutdown)
+{
+	UtilityThreadQueueEntry *entry = (UtilityThreadQueueEntry *) mono_lock_free_queue_dequeue (&thread->work_queue);
+	if (!entry)
+		return FALSE;
+	
+	thread->callbacks.command (thread->state_ptr, &entry->payload, at_shutdown);
+	if (entry->response_sem) {
+		*entry->finished = TRUE;
+		mono_os_sem_post (entry->response_sem);
+	}
+
+	mono_thread_hazardous_try_free (entry, free_queue_entry);
+
+	return TRUE;
+}
+
+static void *
+utility_thread (void *arg)
+{
+	MonoUtilityThread *thread = (MonoUtilityThread *) arg;
+	if (thread->callbacks.early_init)
+		thread->callbacks.early_init (&thread->state_ptr);
+
+	mono_thread_info_wait_inited ();
+	mono_thread_info_attach ();
+
+	thread->callbacks.init (&thread->state_ptr);
+
+	while (mono_atomic_load_i32 (&thread->run_thread)) {
+		MONO_ENTER_GC_SAFE;
```

Comment:
I don't think this is necessary.  And actually may not be right under coop or hybrid suspend - the utility threads is created with `mono_native_thread_create`, so it's not initially attached to the runtime.  That means it doesn't have a `MonoThreadInfo *info`, which means that:
1. the GC doesn't know about it, won't stop it
2. thread state transitions functions will assert because `info` will be `NULL`.

There are a couple of options:
1. just get rid of the thread state transition macros here, and keep using the OS semaphors etc.  The downside is that you have to be vigilant not to call anything from the utility thread that may attach it to the runtime.
2. Create it as an internal thread with `mono_thread_create_internal`, like the profiler does, and set `MONO_THREAD_FLAG_DONT_MANAGE` so that it's not paused by the GC, and then use coop concurrency primitives.

Reply:
The utility thread is attached to the runtime here: https://github.com/mono/mono/pull/9401/files#diff-34d834b8d69724f49f21964be1e1258dR61